src/openrc/rc.c

```diff
@@ -344,22 +347,46 @@ static char *get_krunlevel(void)
 static void
 add_pid(pid_t pid)
 {
+	sigset_t sset, old;
 	RC_PID *p = xmalloc(sizeof(*p));
 	p->pid = pid;
+
+	/* this list will be accessed inside the SIGCHLD signal handler.
+	 * so we need to ensure that the SIGCHLD handler doesn't get invoked
+	 * while the list is at an inconsistent state.
+	 */
+	sigemptyset(&sset);
+	sigaddset(&sset, SIGCHLD);
+	sigprocmask(SIG_SETMASK, &sset, &old);
 	LIST_INSERT_HEAD(&service_pids, p, entries);
+	sigprocmask(SIG_SETMASK, &old, NULL);
 }
 
 static void
-remove_pid(pid_t pid)
+remove_pid(pid_t pid, bool inside_signal)
 {
-	RC_PID *p;
-
-	LIST_FOREACH(p, &service_pids, entries)
-	    if (p->pid == pid) {
-		    LIST_REMOVE(p, entries);
-		    free(p);
-		    return;
-	    }
+	sigset_t sset, old;
+	RC_PID *p, *tmp;
+
+	/* same rationale for blocking SIGCHLD as add_pid() */
+	sigemptyset(&sset);
+	sigaddset(&sset, SIGCHLD);
+	sigprocmask(SIG_SETMASK, &sset, &old);
+	LIST_FOREACH(p, &service_pids, entries) {
+		if (p->pid == pid) {
+			LIST_REMOVE(p, entries);
+			LIST_INSERT_HEAD(&free_these_pids, p, entries);
+			break;
+		}
+	}
+	sigprocmask(SIG_SETMASK, &old, NULL);
+	/* only call free if we're not inside a signal handler */
+	if (!inside_signal) {
+		LIST_FOREACH_SAFE(p, &free_these_pids, entries, tmp) {
+			LIST_REMOVE(p, entries);
```

Comment:
fwiw strictly speaking that list remove ouside of signal is still racy with the list_insert_head inside signal above... (This is much less likely than tripping on another malloc though so I probably won't trigger any issue on this)

While using the "safe" iterator ism't bad per se I think it gives a false impression of security when there isn't any; it feels simpler to just keep the mask a bit longer.
the safe iterators only protect from our own removes, not even concurrent removes (the 'tmp' link could be removed while we hold it, or in the middle of an add to that link), so it's really not that safe at all.

Reply:
> not even concurrent removes

I don't believe there's any concurrent removes going on in the `free_these_pids` list. Unless you were just saying that in general.

> fwiw strictly speaking that list remove ouside of signal is still racy with the list_insert_head inside signal above...

Hmmm, *this* on the other hand is something I can see happening. I guess we'll just have to block `SIGCHLD` for the entire duration of this function.

Reply:
yes, in general, just rambling _safe isn't a good name for this despite being used everywhere.

Anyway, this looks good to me with signal blocked through; I'll update my running test to this exact patch. Thanks!